server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/NewLinesMeasureStep.java

```diff
@@ -0,0 +1,185 @@
+    this.scmInfoRepository = scmInfoRepository;
+  }
+
+  @Override
+  public String getDescription() {
+    return "Compute new lines of code";
+  }
+
+  @Override
+  public void execute() {
+    new PathAwareCrawler<>(
+      FormulaExecutorComponentVisitor.newBuilder(metricRepository, measureRepository)
+        .withVariationSupport(periodsHolder)
+        .buildFor(ImmutableList.of(new NewLinesFormula(measureRepository, scmInfoRepository, metricRepository.getByKey(NCLOC_DATA_KEY)))))
+          .visit(treeRootHolder.getRoot());
+  }
+
+  private static class NewLinesCounter implements Counter<NewLinesCounter> {
+    private final MeasureRepository measureRepository;
+    private final ScmInfoRepository scmInfoRepository;
+    private final Metric nclocDataMetric;
+
+    private final IntVariationValue.Array newLines = IntVariationValue.newArray();
+
+    private NewLinesCounter(MeasureRepository measureRepository, ScmInfoRepository scmInfoRepository, Metric nclocDataMetric) {
+      this.measureRepository = measureRepository;
+      this.scmInfoRepository = scmInfoRepository;
+      this.nclocDataMetric = nclocDataMetric;
+    }
+
+    @Override
+    public void aggregate(NewLinesCounter counter) {
+      this.newLines.incrementAll(counter.newLines);
+    }
+
+    @Override
+    public void initialize(CounterInitializationContext context) {
+      context.getPeriods().forEach(period -> newLines.increment(period, 0));
+
+      Component leak = context.getLeaf();
+      if (leak.getType() != Component.Type.FILE) {
+        return;
+      }
+
+      Optional<ScmInfo> optionalScmInfo = scmInfoRepository.getScmInfo(leak);
+      Optional<Measure> nclocData = measureRepository.getRawMeasure(leak, nclocDataMetric);
+
+      if (!nclocData.isPresent() || !optionalScmInfo.isPresent()) {
+        return;
+      }
+
+      ScmInfo scmInfo = optionalScmInfo.get();
+
+      nclocLineNumbers(nclocData.get()).stream()
+        .map(scmInfo::getChangesetForLine)
+        .forEach(changeset -> context.getPeriods().stream()
+          .filter(period -> isLineInPeriod(changeset, period))
+          .forEach(period -> newLines.increment(period, 1)));
+    }
+
+    /**
+     * NCLOC_DATA contains Key-value pairs, where key - is a line number, and value - is an indicator of whether line
+     * contains code (1) or not (0).
+     *
+     * This method parses the value of the NCLOC_DATA measure and return the line numbers which contain code.
+     */
+    private static List<Integer> nclocLineNumbers(Measure nclocDataMeasure) {
+      Map<Integer, Integer> parsedNclocData = KeyValueFormat.parse(nclocDataMeasure.getData(), newIntegerConverter(), newIntegerConverter());
+      return parsedNclocData.entrySet()
+        .stream()
+        .filter(entry -> entry.getValue() == 1)
+        .map(Map.Entry::getKey)
+        .collect(Collectors.toList());
+    }
+
+    private static boolean isLineInPeriod(Changeset changeset, Period period) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused private "isLineInPeriod" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)